README.md
# ssh-tunnel
Establishes ssh tunnel for forwarding remote port to local one.

Forked from https://github.com/gemini-testing/ssh-tun


CHANGELOG.md
# Changelog

## 1.7.0 - 2019.06.07

* Add option for passing custom SSH arguments

## 1.4.0 - 2017.06.27

* Display a kill signal with which tunnel is closed

## 1.3.0 - 2016.12.12

* Ability to create ssh tunnel via the specified port
* Add `user` public option

## 1.2.0 - 2016.09.30

* Add `host` & `port` public options

## 1.1.0 - 2016.08.31

* Add `user` option

## 1.0.0 - 2016.02.18

* Initial release


test/index.js
var Tunnel = require('../'),
    _ = require('lodash'),
    q = require('q'),
    childProcess = require('child_process'),
    events = require('events'),
    util = require('util');

describe('Tunnel', function () {
    var sandbox = sinon.sandbox.create(),
        tunnel,
        ssh;

    beforeEach(function () {
        ssh = new events.EventEmitter();
        ssh.stderr = new events.EventEmitter();
        ssh.kill = sandbox.stub();

        sandbox.stub(console, 'info');
        sandbox.stub(childProcess);
        childProcess.spawn.returns(ssh);
    });

    afterEach(function () {
        sandbox.restore();
    });

    describe('constructor', function () {
        it('should generate proxy host from host and ports range', function () {
            var tunnel = createTunnel({
                host: 'test_host',
                ports: { min: 0, max: 1 }
            });

            expect(tunnel.proxyHost).to.match(/test_host:\d+/);
        });

        it('should set port as passed in ports if min and max ports are same', function () {
            var tunnel = createTunnel({
                ports: { min: 8080, max: 8080 }
            });

            expect(tunnel.proxyHost).to.have.string(':8080');
        });

        it('should set default timeout as 10 seconds', function () {
            var tunnel = createTunnel(),
                timeout = sandbox.stub(q.makePromise.prototype, 'timeout'); // promise constructor available like this?!

            tunnel.open();

            expect(timeout).to.be.calledWith(10000);
        });
    });

    describe('instance methods', function () {
        describe('open', function () {
            it('should return promise', function () {
                tunnel = createTunnel();

                var result = tunnel.open();

                expect(q.isPromise(result)).to.be.true;
            });

            describe('tunnel spawn and events', function () {
                it('should spawn ssh process', function () {
                    tunnel = createTunnel();

                    tunnel.open();

                    expect(childProcess.spawn).to.be.calledWith('ssh');
                });

                it('should spawn tunnel from localhost to remote host for provided local and remote ' +
                    'ports', function () {
                    tunnel = createTunnel({
                        host: 'remote_host',
                        localport: 8080
                    });

                    tunnel.open();

                    var sshArgs = childProcess.spawn.lastCall.args[1];

                    expect(sshArgs[0]).to.match(/-R \d+:localhost:8080/);
                    expect(sshArgs).to.contain('remote_host')
                        .and.to.contain('-N')
                        .and.to.contain('-v');
                });

                it('should spawn tunnel via the default ssh port', function () {
                    createTunnel().open();

                    var sshArgs = childProcess.spawn.lastCall.args[1];

                    expect(sshArgs).to.contain('-p 22');
                });

                it('should spawn tunnel via the specified ssh port', function () {
                    createTunnel({ sshPort: 100500 }).open();

                    var sshArgs = childProcess.spawn.lastCall.args[1];

                    expect(sshArgs).to.contain('-p 100500');
                });

                it('should spawn tunnel to remote host using provided user ', function () {
                    tunnel = createTunnel({
                        host: 'remote_host',
                        user: 'user',
                        localport: 8080
                    });

                    tunnel.open();

                    var sshArgs = childProcess.spawn.lastCall.args[1];

                    expect(sshArgs).to.contain('user@remote_host');
                });

                it('should spawn tunnel with StrictHostKeyChecking=no if host verification is disabled', function () {
                    tunnel = createTunnel({
                        strictHostKeyChecking: false
                    });

                    tunnel.open();

                    var sshArgs = childProcess.spawn.lastCall.args[1];

                    expect(sshArgs).to.contain('-o StrictHostKeyChecking=no');
                });

                it('should spawn tunnel with provided identity file disabled', function () {
                    tunnel = createTunnel({
                        identity: '~/.ssh/id_rsa_alt'
                    });

                    tunnel.open();

                    var sshArgs = childProcess.spawn.lastCall.args[1];

                    expect(sshArgs).to.contain('-i ~/.ssh/id_rsa_alt');
                });

                it('should spawn tunnel with provided additional SSH arguments', function () {
                    tunnel = createTunnel({
                        additionalSSHArgs: ['-o ControlPath=no']
                    });

                    tunnel.open();

                    var sshArgs = childProcess.spawn.lastCall.args[1];

                    expect(sshArgs).to.contain('-o ControlPath=no');
                });

                it('should resolve promise if tunnel successfully created', function () {
                    tunnel = createTunnel();

                    var promise = tunnel.open();

                    ssh.stderr.emit('data', 'success');

                    return expect(promise).to.be.eventually.resolved;
                });

                it('should set resulting host and port to public variable', function () {
                    tunnel = createTunnel({
                        host: 'some_ssh_host'
                    });

                    tunnel.open();

                    ssh.stderr.emit('data', 'success');

                    expect(tunnel.proxyHost).to.match(/^some_ssh_host:\d+/);
                });

                it('should reject tunnel opening if failed to create tunnel', function () {
                    tunnel = createTunnel();

                    var openPromise = tunnel.open();

                    ssh.stderr.emit('data', 'failed');

                    expect(openPromise).to.be.rejected;
                });

                it('should do nothing if fail occured in stderr after successful tunnel opening', function () {
                    tunnel = createTunnel();

                    tunnel.open();
                    ssh.stderr.emit('data', 'success');
                    ssh.stderr.emit('data', 'failed');

                    expect(console.info)
                        .to.be.not.calledWith(util.format('ERROR: failed to create tunnel to %s.', tunnel.proxyHost));
                });

                it('should reject tunnel opening if error occured', function () {
                    tunnel = createTunnel();

                    var openPromise = tunnel.open();

                    ssh.emit('error');

                    expect(openPromise).to.be.rejected;
                });

                it('should close tunnel if close event received', function () {
                    tunnel = createTunnel();
                    tunnel.open();

                    var closePromise = tunnel.close();

                    ssh.emit('close');

                    return expect(closePromise).to.be.eventually.resolved;
                });
            });
        });

        describe('close', function () {
            it('should return resolved promise if no tunnel opened', function () {
                tunnel = createTunnel();

                expect(tunnel.close()).to.eventually.resolved;
            });

            it('should return promise if tunnel opened', function () {
                tunnel = createTunnel();
                tunnel.open();

                expect(q.isPromise(tunnel.close())).to.be.true;
            });

            it('should try to kill tunnel using SIGTERM', function () {
                tunnel = createTunnel();
                tunnel.open();
                tunnel.close();

                expect(ssh.kill).to.be.calledWith('SIGTERM');
                expect(ssh.kill).to.be.not.calledWith('SIGKILL');
            });

            it('should try to kill tunnel with SIGKILL if tunnel was not closed in 3000ms after ' +
                'SIGTERM', function () {
                var clock = sinon.useFakeTimers(),
                    closePromise;

                tunnel = createTunnel();
                tunnel.open();

                closePromise = tunnel.close();
                expect(ssh.kill).to.be.calledWith('SIGTERM');

                clock.tick(3100);

                return closePromise.then(function () {
                    expect(ssh.kill).to.be.calledWith('SIGKILL');
                    clock.restore();
                });
            });

            it('should emit close event', function () {
                tunnel = createTunnel();
                tunnel.open();

                var closeHandler = sandbox.stub();
                tunnel.on('close', closeHandler);
                ssh.emit('close', null, 'SIGKILL');

                return expect(closeHandler).to.be.calledWith(null, 'SIGKILL');
            });

            it('should emit exit event', function () {
                tunnel = createTunnel();
                tunnel.open();

                var exitHandler = sandbox.stub();
                tunnel.on('exit', exitHandler);
                ssh.emit('exit', 0, null);

                return expect(exitHandler).to.be.calledWith(0, null);
            });
        });
    });

    describe('static methods', function () {
        describe('openWithRetries', function () {
            beforeEach(function () {
                sandbox.stub(Tunnel.prototype);
                Tunnel.prototype.open.returns(q());
                Tunnel.prototype.close.returns(q());
            });

            it('should return promise', function () {
                var result = Tunnel.openWithRetries(defaultOpts());

                expect(q.isPromise(result)).to.be.true;
            });

            it('should try to open tunnel with passed opts', function () {
                var opts = defaultOpts();

                Tunnel.openWithRetries(opts);

                expect(Tunnel.prototype.__constructor).to.be.calledWith(opts);
            });

            it('should resolve promise if tunnel opened successfully', function () {
                Tunnel.prototype.open.returns(q());

                return expect(Tunnel.openWithRetries(defaultOpts())).to.be.eventually.resolved;
            });

            it('should resolve promise with tunnel instance', function () {
                Tunnel.prototype.open.returns(q());

                return Tunnel.openWithRetries(defaultOpts()).then(function (tunnel) {
                    expect(tunnel).to.be.instanceOf(Tunnel);
                });
            });

            it('should reject tunnel if failed to open tunnel after retries', function () {
                Tunnel.prototype.open.returns(q.reject());

                return expect(Tunnel.openWithRetries(defaultOpts)).to.be.eventually.rejected;
            });

            it('should retry to create tunnel 5 times by default', function () {
                Tunnel.prototype.open.returns(q.reject());

                return Tunnel.openWithRetries(defaultOpts()).fail(function () {
                    expect(Tunnel.prototype.open.callCount).to.be.equal(5);
                });
            });

            it('should retry create tunnel retries times', function () {
                Tunnel.prototype.open.returns(q.reject());

                return Tunnel.openWithRetries(defaultOpts(), 10).fail(function () {
                    expect(Tunnel.prototype.open.callCount).to.be.equal(10);
                });
            });
        });
    });
});

function createTunnel (opts) {
    opts = opts || {};

    _.defaults(opts, defaultOpts());

    return new Tunnel(opts);
}

function defaultOpts () {
    return {
        host: 'deafault_host',
        ports: { min: 0, max: 1 },
        localport: 0
    };
}


index.js
'use strict';

var inherit = require('inherit'),
    q = require('q'),
    childProcess = require('child_process'),
    util = require('util'),
    EventEmitter = require('events').EventEmitter;

var DEFAULTS = {
    MAX_RETRIES: 5,
    CONNECT_TIMEOUT: 10000,
    SSH_PORT: 22
};

var Tunnel = inherit(EventEmitter, {
    /**
     * Constuctor
     * @param {object} opts tunnel options
     * @param {string} opts.host remote host address
     * @param {object} opts.ports remote host ports range
     * @param {number} opts.localport local port number
     * @param {string} [opts.user] remote host user
     * @param {string} [opts.sshPort=22] host port to create tunnel
     * @param {number} [opts.maxRetries=5] max attempts to create tunnel
     * @param {number} [opts.connectTimeout=10000] ssh connect timeout
     * @param {boolean} [opts.strictHostKeyChecking=true] verify host authenticity
     * @param {string} [opts.identity] private key for public key authentication
     * @param {string[]} [opts.additionalSSHArgs] additional args for SSH command
     */
    __constructor: function (opts) {
        EventEmitter.call(this);

        this.host = opts.host;
        this.port = this._generateRandomPort(opts.ports);
        this._sshPort = opts.sshPort || DEFAULTS.SSH_PORT;
        this.user = opts.user;
        this.proxyHost = util.format('%s:%d', this.host, this.port);
        this.proxyUrl = this.proxyHost; // deprecated, use proxyHost
        this._localPort = opts.localport;
        this._connectTimeout = opts.connectTimeout || DEFAULTS.CONNECT_TIMEOUT;
        this._tunnel = null;
        this._tunnelDeferred = q.defer();
        this._closeDeferred = q.defer();
        this._strictHostKeyChecking = opts.strictHostKeyChecking === undefined ? true : opts.strictHostKeyChecking;
        this._identity = opts.identity;
        this._additionalSSHArgs = opts.additionalSSHArgs || [];
    },

    /**
     * Tries to open ssh connection to remote server
     * @returns {Promise}
     */
    open: function () {
        var _this = this;

        console.info('INFO: creating tunnel to %s', this.proxyHost);

        this._tunnel = childProcess.spawn('ssh', this._buildSSHArgs());

        var cleanup = function () {
            _this._tunnel.stderr.removeAllListeners('data');
        };

        this._tunnel.stderr.on('data', function (data) {
            if (/success/.test(data)) {
                cleanup();
                return _this._resolveTunnel();
            }

            if (/failed/.test(data)) {
                cleanup();
                return _this._rejectTunnel();
            }
        });

        this._tunnel.on('exit', function (code, signal) {
            _this.emit('exit', code, signal);
        });

        this._tunnel.on('close', function (code, signal) {
            _this.emit('close', code, signal);
            return _this._closeTunnel(code);
        });

        this._tunnel.on('error', function () {
            return _this._rejectTunnel();
        });

        return _this._tunnelDeferred.promise.timeout(this._connectTimeout);
    },

    /**
     * Closes connection. If no connection established does nothing
     * @returns {Promise}
     */
    close: function () {
        if (!this._tunnel) {
            return q();
        }

        var _this = this;

        this._tunnel.kill('SIGTERM');
        return this._closeDeferred.promise.timeout(3000).fail(function () {
            _this._tunnel.kill('SIGKILL');
            return _this._closeTunnel(-1);
        });
    },

    _resolveTunnel: function () {
        console.info('INFO: Tunnel created to %s', this.proxyHost);
        this._tunnelDeferred.resolve();
    },

    _rejectTunnel: function () {
        var message = util.format('ERROR: failed to create tunnel to %s.', this.proxyHost),
            error = new Error(message);

        console.info(message);
        this._tunnelDeferred.reject(error);
    },

    _closeTunnel: function (exitCode) {
        console.info('INFO: Tunnel to %s closed. Exit code: %d', this.proxyHost, exitCode);
        this._closeDeferred.resolve();
    },

    _buildSSHArgs: function () {
        return [
            util.format('-R %d:localhost:%d', this.port, this._localPort),
            '-N',
            '-v',
            this._strictHostKeyChecking === false ? '-o StrictHostKeyChecking=no' : '',
            this._identity ? util.format('-i %s', this._identity) : '',
            util.format('-p %d', this._sshPort),
            (this.user ? this.user + '@' : '') + this.host
        ].concat(this._additionalSSHArgs).filter(Boolean);
    },

    _generateRandomPort: function (ports) {
        var min = ports.min,
            max = ports.max;

        return Math.floor(Math.random() * (max - min + 1)) + min;
    }
});

/**
 * Tries to open tunnel several times
 * @param {object} opts opts which will be passed to created tunnel
 * @param {number} retries amount of retries to open tunnel
 * @returns {Promise}
 */
Tunnel.openWithRetries = function (opts, retries) {
    retries = retries || DEFAULTS.MAX_RETRIES;

    function retry_(retriesLeft) {
        if (!retriesLeft) {
            return q.reject(util.format('ERROR: failed to create tunnel after %d attempts', retries));
        }

        var tunnel = new Tunnel(opts);

        return tunnel.open()
            .then(function () {
                return q.resolve(tunnel);
            })
            .fail(function () {
                return tunnel.close()
                    .then(retry_.bind(null, retriesLeft - 1));
            });
    }

    return retry_(retries);
};

module.exports = Tunnel;
